l5kit/l5kit/environment/envs/l5_env.py

```diff
@@ -0,0 +1,200 @@
+        :return: the observation of first frame of sampled scene index
+        """
+
+        # Sample a episode randomly
+        self.scene_indices = [random.randint(0, self.max_scene_id)]
+        self.sim_dataset = SimulationDataset.from_dataset_indices(self.dataset, self.scene_indices, self.sim_cfg)
+
+        # Define in / outs for given scene
+        self.agents_ins_outs: DefaultDict[int, List[List[UnrollInputOutput]]] = defaultdict(list)
+        self.ego_ins_outs: DefaultDict[int, List[UnrollInputOutput]] = defaultdict(list)
+
+        # Reset CLE evaluator
+        self.reward.reset()
+
+        # Output first observation
+        self.frame_index = 0
+        ego_input = self.sim_dataset.rasterise_frame_batch(self.frame_index)
+        self.ego_input_dict = default_collate_numpy(ego_input[0])
+
+        # Only output the image attribute
+        obs = {'image': ego_input[0]["image"]}
+        return obs
+
+    def step(self, action: np.ndarray) -> GymStepOutput:
+        """
+        Analogous to sim_loop.unroll(scenes_to_unroll) in L5Kit
+
+        :param action: the action to perform on current state/frame
+        :return: the namedTuple comprising the (next observation, reward, done, info)
+                 based on the current action
+        """
+
+        frame_index = self.frame_index
+        next_frame_index = frame_index + 1
+        episode_over = next_frame_index == (len(self.sim_dataset) - self.future_num_frames)  # Note !!
+
+        # EGO
+        if not self.sim_cfg.use_ego_gt:
+            action = convert_to_dict(action, self.future_num_frames)
+            self.ego_output_dict = action
+
+            if not self.open_loop:
+                # Update raster according to predicted ego since closed loop training
+                self.simulator.update_ego(self.sim_dataset, next_frame_index, self.ego_input_dict, self.ego_output_dict)
+
+            ego_frame_in_out = self.simulator.get_ego_in_out(self.ego_input_dict, self.ego_output_dict,
+                                                             self.simulator.keys_to_exclude)
+            for scene_idx in self.scene_indices:
+                self.ego_ins_outs[scene_idx].append(ego_frame_in_out[scene_idx])
+
+        # reward calculation
+        if self.open_loop:
+            reward = self.reward.get_open_loop_reward(self.ego_output_dict, self.ego_input_dict)
+        else:
+            reward = self.reward.get_close_loop_reward(self.frame_index, self.scene_indices, self.sim_dataset,
+                                                       self.ego_ins_outs, self.agents_ins_outs)
+
+        # done is True when episode ends or error gets too high (optional)
+        done = episode_over or self.check_done_status()
+
+        # Optionally we can pass additional info
+        # We are using "info" to output simulated outputs
+        info = {}
+        if done:
+            info = {"info": self.get_simulated_outputs()}
```

Comment:
Do we need this ? This will be copied through IPC so it might be very expensive.